### ISSUE TITLE
Question related to Input Preparation 

### ISSUE THREAD

**Issue by XuyangBai** [HuguesTHOMAS/KPConv#12]
Hello, thank you for your sharing, I really appreciate your KPConv work and I am working on implement it in PyTorch.  But I was confused when I try to understand the code of data preparation part because there are so many encapsulations or wrappers. So would you please give me some high level idea about the implementation. Basically my questions are 

1. About the `flat_inputs` variable : In my understanding, you build the `flat_inputs` which consists of the points and its neighbors for each layer before session run. But how can you calculate the neighbors of each layer before the specific point cloud is sent into the network? 
2. For the convolutional part, I already know how to build the matrix and calculate the output feature for the batch input, but before conv I need know the neighborhood indices of each point, I have no idea how to calculate the neighbor for each point cloud in parallel, obviously if I try to calculate the neighbor indices one by one and then concatenation it will be very time comsuing. So in your implementation, how do you solve this problem ?

**Comment by HuguesTHOMAS** [comment]
Hello @XuyangBai,

Thank you for your interest, a PyTorch implementation is a great idea, unfortunately I did not have any time to do it, but I can help you for sure.

The funny thing is that your second question is the answer to your first question and your first question is... Well, the answer of your second question. Here is my line of thoughts when I implemented the whole thing. 

First, as you noticed, we need the matrix of neighbors in the convolution operator, and finding neighbors in a point cloud is not a task that you can easily implement on a GPU. In fact it is quite well optimized for CPU, with the used of KD-Tree to accelerate the search time. Therefore, I decided to pre-compute the neighbors, before sending them to the GPU. It is the same for pooling indices, which are the point neighbors from one layer to the other. We thus need the points of each layer beforehand, which is possible because our pooling operation is purely geometric (we do not use the features of the network to adapt the pooling location). 

In brief, there are two groups of operations:

* The input pipeline: generate the input point clouds, process them with data augmentation, subsample for every layer, find neighbors/pooling indices. Everything here is done on CPU, with an parallel input queue (8 threads precomputing a queue of input batches).

* The network graph: all the layer ops in GPU. The network takes the first batch from the input queue , process it, etc.

Tensorflow already have a `Dataset` class that implements parallel input queues with multiple threads, but it need every operation to be a tensorflow op. This is why I implemented 2 tensorflow C++ wrappers in the folder `tf_custom_op`, one for neighbor search (with the [nanoflann](https://github.com/jlblancoc/nanoflann) library), and one for subsampling.

The python C++ wrappers in the folder `cpp_wrappers` are not used by the network or the input pipeline, they only get called at preprocessing steps (the part of the code that run only once the first time your start a training). Actually there is only one subsampling wrapper, and it can be replaced by a python function if you wish.

The `flat_inputs` variable is a list of all the input tensors, returned by the input pipeline. It contains a lot of things, with some variables that are not really necessary, but that I used for debugging. For convenience, I convert it to a dictionary afterwards in the model class.

I hope my explanations are clear enough. If you have any other question, feel free to ask. Did you catch the trick to manage the shadow neighbors? And the bacth stacking process, with the batch length calibration? Another part of the code that is not explained in the paper is the neighborhood calibration function, which is quite important for speed.

Best,
Hugues

**Comment by amiltonwong** [comment]
@HuguesTHOMAS ,
Could elaborate a bit on neighborhood calibration function, which is found in [common.py](https://github.com/HuguesTHOMAS/KPConv/blob/master/datasets/common.py)? Why using it and how to perform it?

THX!

**Comment by XuyangBai** [comment]
Hi @HuguesTHOMAS 

Thank you so much for your clear explanation. Now I understand how you prepare the neighbor indices and the function of `tf_custom_op`.

For my current implementation, I just downsample the point cloud into the same number of points, so it is quite easy for me to handle the shadow neighbors and also the number of point cloud fragment of one batch is fixed. For the batch stacking process, I think you just end the loop when the total number of points reach a certain upper bound?  But I am not sure what the neighborhood calibration function for, either. Could you please give some explanation.

**Comment by HuguesTHOMAS** [comment]
@XuyangBai,
I know this is much easier to implement, but I would not advise you to downsample with the same number of points every time. The whole point of this method it to be geometrically consistent. You have to control the density of points rigorously, with grid subsampling.

@amiltonwong and @XuyangBai,
The neighborhood calibration function is here to control the number of neighbors. As we preferred radius neighborhood (for geometrical consistency), some of the neighborhoods (in the dense areas like vegetation) can contain a lot of points. For an input batch, the neighborhood matrix size is `[N, n_max]`  where `n_max` is the maximum number of neighbors. As we have big batches, there is often a dense area among the points, forcing `n_max` to be large.  Furthermore, in the case the density is not controlled with a grid, we could end up with very large `n_max`, causing an OOM crash. The neighborhood calibration function sets a limit for `n_max`, by checking some input batches. The limit (set for each layer of the network in the `self.neighborhood_limits` variable of the `dataset` class) is set as the 80th percentile of the distribution of neighbor numbers. It means that 80% of the neighborhoods won't be affected by this limit, and that the 20% most dense neighborhoods will lose some of their points.  Because of the way we compute neighborhoods, they lose the furthest points, which means they become KNN neighborhoods.
Thanks to this trick, our network is lighter, faster, and it does not affect the performances

**Comment by lelouedec** [comment]
@XuyangBai  Any news concerning your implementation using pytorch ? I am considering doing the same thing to use and benchmark KPConv for my research.

**Comment by XuyangBai** [comment]
@lelouedec Hi,  actually I am busy with other deadlines recently. I have implement the basic convolution operation, but I still need to optimize my code to make it faster because my current version is too slow. I can share my code with you if you wish.

**Comment by tchaton** [comment]
@amiltonwong @lelouedec @XuyangBai @HuguesTHOMAS  Any progress on KPconv implementation in Pytorch. Pytorch geometric would be a great place to make the implémentation if you guys can. I might work on it too.

**Comment by XuyangBai** [comment]
Hi all, I have just finished a initial version of PyTorch implementation [here](https://github.com/XuyangBai/KPConv.pytorch).
